### PR TITLE
Support custom http headers as a CouchConfig option

### PR DESCRIPTION
**Pull request by tomblench**

### CODE REVIEW

File: sync-core/src/main/java/com/cloudant/mazha/CouchConfig.java [hunk 1]
```diff
@@ -144,6 +151,23 @@ public void setMaxConnections(int maxConnections) {
...
+        return customHeaders;
+    }
+
+    public void setCustomHeaders(Map<String, String> customHeaders) {
+        this.customHeaders = new HashMap<String, String>();
+        List<String> prohibitedHeaders = Arrays.asList("authorization", "www-authenticate",
+                "host", "connection", "content-type", "accept", "content-length");
+        for (Map.Entry<String, String> header : customHeaders.entrySet()) {
+            if (prohibitedHeaders.contains(header.getKey().toLowerCase())) {
+                throw new IllegalArgumentException("Bad optional HTTP header: " + header);
```
Comment (rhyshort) [inline sync-core/src/main/java/com/cloudant/mazha/CouchConfig.java]: I think we should have a better exception message, it should state that the header is not allowed to be set by the user

File: sync-core/src/test/java/com/cloudant/mazha/HttpRequestsTest.java [hunk 2]
```diff
@@ -57,6 +59,37 @@ public void get_resourceExist_success() throws Exception {
         Assert.assertTrue(s.contains(TEST_DB));
     }
 
+    @Test(expected = IllegalArgumentException.class)
+    public void prohibitedCustomHeaderFails() throws Exception {
+        CouchConfig customCouchConfig = getCouchConfig();
+        Map<String, String> customHeaders = new HashMap<String, String>();
+        customHeaders.put("WWW-Authenticate", "test");
```
Comment (mikerhodes) [inline sync-core/src/test/java/com/cloudant/mazha/HttpRequestsTest.java]: This should be a loop with all the prohibited ones.

File: sync-core/src/main/java/com/cloudant/mazha/HttpRequests.java [hunk 3]
```diff
@@ -110,11 +112,29 @@ public int getMaxForRoute(HttpRoute route) {
...
+                }
+            } else {
+                if (authHeaderValue != null) {
+                    // if we're doing basic auth, make a copy of the custom headers and add it in
+                    this.requestHeaders = new HashMap<String, String>();
+                    customHeaders.putAll(customHeaders);
+                    customHeaders.put("authorization", authHeaderValue);
+                }
+            }
+        }
```
Comment (mikerhodes) [inline sync-core/src/main/java/com/cloudant/mazha/HttpRequests.java]: It looks like here the Authorization header will not be set if there are no already-existing custom headers (`customHeaders != null`).

**Comment by mikerhodes** [comment]
Looks good, squash and merge.

**Comment by tomblench** [comment]
Fixes #51

**Pull request merged**